### PR TITLE
 App update validator removed

### PR DESCRIPTION
**Pull request by meln1k**
Summary: App update validation process is simplified now: instead of validation app update and the resulting app definition, we merge the update first and validate the result. Note that app normalization stays there as before.

Jira issues: MARATHON-8498

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala [hunk 1]
```diff
@@ -22,15 +22,12 @@ object AppHelpers {
...
-    enabledFeatures: Set[String], config: AppNormalization.Config): NormalizationWithContext[raml.AppUpdate, Map[String, SecretDef]] =
-    Normalization.withContext {
-      case (app, existingSecrets) =>
-        validateOrThrow(app)(AppValidation.validateOldAppUpdateAPI)
-        val migrated = AppNormalization.forDeprecatedUpdates(config).normalized(app)
-        validateOrThrow(app)(AppValidation.validateCanonicalAppUpdateAPI(enabledFeatures, () => config.defaultNetworkName, existingSecrets.getOrElse(Map.empty)))
-        AppNormalization.forUpdates(config).normalized(migrated)
-    }
+  def appUpdateNormalization(config: AppNormalization.Config): Normalization[raml.AppUpdate] = Normalization { app =>
+    //            validateOrThrow(app)(AppValidation.validateOldAppUpdateAPI)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala]: This should go, right?

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala [hunk 2]
```diff
@@ -333,12 +335,12 @@ class AppsResource @Inject() (
 
     val updateType = (appDoesNotExist, partialUpdate) match {
       case (true, _) => CompleteReplacement
-      case (_, true) => PartialUpdate
+      case (_, true) => PartialUpdate(maybeExistingApp.getOrElse(throw new RuntimeException("This should never happen")))
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: The exception message could be a little more explicit about what happened.

File: src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateFormatTest.scala [hunk 3]
```diff
@@ -45,13 +37,6 @@ class AppUpdateFormatTest extends UnitTest {
       appUpdate.acceptedResourceRoles should equal(Some(Set(ResourceRole.Unreserved)))
     }
 
-    "FromJSON should fail when 'acceptedResourceRoles' is defined but empty" in {
-      val json = """ { "id": "test", "acceptedResourceRoles": [] }"""
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateFormatTest.scala]: Is this still true for our update? Does the validator fail when we merge this update with an app?

File: src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateTest.scala [hunk 4]
```diff
@@ -210,53 +170,6 @@ class AppUpdateTest extends UnitTest with ValidationTestLike {
       assert(changed == app.copy(acceptedResourceRoles = Set("b")))
     }
 
-    "AppUpdate with a version and other changes are not allowed" in {
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateTest.scala]: I think we have to make sure that there are still validations error with such updates. See my comment above.

File: src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateFormatTest.scala (hunk 3 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateFormatTest.scala]: Yes, check this line:

https://github.com/mesosphere/marathon/blob/master/src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionFormatsTest.scala#L213

File: src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala [hunk 5]
```diff
@@ -22,15 +22,12 @@ object AppHelpers {
...
+    val migrated = AppNormalization.forDeprecatedUpdates(config).normalized(app)
+    //            validateOrThrow(app)(AppValidation.validateCanonicalAppUpdateAPI(enabledFeatures, () => config.defaultNetworkName, existingSecrets.getOrElse(Map.empty)))
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala]: Why was this disabled?

File: src/test/scala/mesosphere/marathon/api/validation/AppUpdateValidatorTest.scala [hunk 6]
```diff
@@ -1,133 +0,0 @@
...
-import mesosphere.marathon.raml.{App, AppCContainer, AppUpdate, ContainerPortMapping, EngineType, Raml}
-import mesosphere.marathon.state.AppDefinition
-import org.scalatest.Matchers
-import play.api.libs.json.Json
-
-class AppUpdateValidatorTest extends UnitTest with Matchers {
-  implicit val appUpdateValidator = AppValidation.validateCanonicalAppUpdateAPI(Set.empty, () => AppNormalization.Configuration(None, "mesos-bridge-name").defaultNetworkName, Map.empty)
-  implicit val validAppDefinition = AppDefinition.validAppDefinition(Set.empty)(PluginManager.None)
-
-  "validation that considers container types" should {
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/api/validation/AppUpdateValidatorTest.scala]: What about these tests (and the ones above)? Are *all* those cases properly covered in other tests?

File: src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateTest.scala (hunk 4 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateTest.scala]: Good catch! I added a test for this. I checked other parts too and it seems they cover everything.

File: src/test/scala/mesosphere/marathon/api/v2/AppsResourceTest.scala [hunk 7]
```diff
@@ -1639,6 +1638,28 @@ class AppsResourceTest extends AkkaUnitTest with GroupCreation with JerseyTest {
...
+        cmd = Some("cmd"),
+        instances = 1
+      )
+      prepareApp(app, groupManager) // app is stored
+
+      When("The application is updated")
+      val updateRequest = App(id = id, instances = 2, version = Some(OffsetDateTime.now()))
+      val updatedBody = Json.stringify(Json.toJson(updateRequest)).getBytes("UTF-8")
+      val response = asyncRequest { r =>
+        appsResource.patch(app.id, updatedBody, force = false, auth.request, r)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/v2/AppsResourceTest.scala]: What is the reason for testing this not in `AppValidationTest`?

File: src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala [hunk 8]
```diff
@@ -22,15 +22,11 @@ object AppHelpers {
...
-    Normalization.withContext {
-      case (app, existingSecrets) =>
-        validateOrThrow(app)(AppValidation.validateOldAppUpdateAPI)
-        val migrated = AppNormalization.forDeprecatedUpdates(config).normalized(app)
-        validateOrThrow(app)(AppValidation.validateCanonicalAppUpdateAPI(enabledFeatures, () => config.defaultNetworkName, existingSecrets.getOrElse(Map.empty)))
-        AppNormalization.forUpdates(config).normalized(migrated)
-    }
+  def appUpdateNormalization(config: AppNormalization.Config): Normalization[raml.AppUpdate] = Normalization { app =>
+    val migrated = AppNormalization.forDeprecatedUpdates(config).normalized(app)
+    validateOrThrow(migrated)(AppValidation.validateAppUpdateVersion)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala]: Ah, so we do have validations that we cannot have as part of the whole app.

File: src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateTest.scala (hunk 4 above)
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/api/v2/json/AppUpdateTest.scala]: I can see

1. `assert(vfe.failure.violations.toString.contains("The 'version' field may only be combined with the 'id' field."))`
2. `assert(vfe.failure.violations.toString.contains("may not be set in conjunction with fetch"))`
3. `assert(vfe.failure.violations.toString.contains("cannot specify both ports and port definitions"))`
4. `assert(vfe.failure.violations.toString.contains("ports must be unique"))`

1 is covered in `AppResourceTest`. 4 is covered in `AppValidationTest`. Where do you test 2 and 4? I think we do have the validations but I'd like to confirm this with tests.

File: src/test/scala/mesosphere/marathon/api/validation/AppUpdateValidatorTest.scala (hunk 6 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/api/validation/AppUpdateValidatorTest.scala]: I checked every test to make sure they don't disappear :)

**Pull request merged**